### PR TITLE
Detect and warn about threads without CPU time instrumentation

### PR DESCRIPTION
**Pull request by ivoanjo**
When a thread is missing CPU instrumentation due to it being started
before our `Ext::CThread` was applied (or in a way which skips our
custom thread initialization code), the resulting CPU profile will
be incomplete -- it will be missing the CPU timing data from those
threads.

This means that regarding CPU instrumentation, we can be in one of
three cases:
1. CPU instrumentation is available AND every thread is correctly
   instrumented
2. CPU instrumentation is available BUT there are threads that are
   missing the instrumentation
3. CPU instrumentation is NOT available

Our ideal case is # 1, and then # 3. Being in # 2 is not desirable,
because it means that our profiling data is incomplete, and thus
may lead customers astray or to lose confidence in our results.

This commits adds a detector that distinguishes between case # 1
and # 2, printing a warning so that customers can report back to
us their issue, and will be aware that there can be more than
what they are seeing in our UX.
(Previously, we would silently ignore missing instrumentation).

This detector works by having the Stack collector record -- at
initialization time -- if its own thread is instrumented, because
instrumentation is always applied before starting the collector.

If no instrumentation is available, then the collector is running
in case # 3. If instrumentation is available, the collector
assumes it is running in either case # 1 or # 2.

During execution, if the Stack collector finds a thread without
instrumentation, it it will know that we are in case # 2 above,
and thus will emit a warning for missing data.

Example way of triggering warning:

```
$ DD_PROFILING_ENABLED=true bundle exec ruby -e 'Thread.start { require "ddtrace/profiling/preload"; sleep }; sleep'
W, [2021-01-15T11:15:04.621929 #506]  WARN -- ddtrace: [ddtrace] Detected thread ('#<Thread:0x0000563aa6326158>') with missing CPU profiling instrumentation. CPU Profiling results will be inaccurate. Please report this at https://github.com/DataDog/dd-trace-rb/blob/master/CONTRIBUTING.md#found-a-bug
```

### CODE REVIEW

File: spec/ddtrace/profiling/ext/cthread_spec.rb
Comment (delner) [inline spec/ddtrace/profiling/ext/cthread_spec.rb:26]: Interesting, I like this idea of using `alias_method` for keeping `initialize`.

File: spec/ddtrace/profiling/ext/cthread_spec.rb
Comment (delner) [inline spec/ddtrace/profiling/ext/cthread_spec.rb:40]: I see how this works;what do you think about defining it as a `thread_class_without_instrumentation`?

That way we could compose it in by substituting the class instead of using `before`? Makes it easier to compose and control conditional test branches when there are fewer `before` statements.

**Review by delner — COMMENTED** [review]
Overall I like the simplicity and the directness (use of variable names for clarity.) Just some questions about how/where we implement the check.

File: lib/ddtrace/profiling/collectors/stack.rb
Comment (ivoanjo) [reply lib/ddtrace/profiling/collectors/stack.rb:127]: This being Ruby, I won't claim it's impossible, but I would claim it's very unlikely:

1. We're getting the list of threads from `Thread.list` and thus by definition everything that is there should be an instance of `Thread`.
2. When a module gets included in a class, it affects all instances of the class, even those existing before the method was included:

  ```ruby
[3] pry(main)> class Someclass
[3] pry(main)* end
=> nil
[4] pry(main)> existing = Someclass.new
=> #<Someclass:0x00007f8265a9d4b0>
[5] pry(main)> existing.foo
NoMethodError: undefined method `foo' for #<Someclass:0x00007f8265a9d4b0>
Did you mean?  for
from (pry):4:in `__pry__'
[6] pry(main)> module SomeModule
[6] pry(main)*   def foo
[6] pry(main)*     :foo
[6] pry(main)*   end
[6] pry(main)* end
=> :foo
[7] pry(main)> class Someclass
[7] pry(main)*   include SomeModule
[7] pry(main)* end
=> Someclass
[8] pry(main)> existing.foo
=> :foo
```

So by #1 + #2 I would expect that if _one_ thread instance passed the `cpu_time_expected_to_work` test in the initializer, all of them should as well.

One extra detail is that you can't even un-include cleanly. I needed to get quite creative in the tests to simulate a case where initialize gets broken, exactly because I couldn't find any other way of skipping our module and only calling the regular `initialize`. So I'd be struggling to find a case where not every thread is `kind_of?(CThread)` once one of them is.

What do you think?

File: spec/ddtrace/profiling/ext/cthread_spec.rb
Comment (ivoanjo) [reply spec/ddtrace/profiling/ext/cthread_spec.rb:26]: As I touched on above, it was quite hard to simulate this "I add another module after the cthread module, and it breaks cthread initialization" (which is good news, I guess). In particular, if you create the alias AFTER including cthread, the alias will include the cthread initialize.

File: spec/ddtrace/profiling/ext/cthread_spec.rb
Comment (ivoanjo) [reply spec/ddtrace/profiling/ext/cthread_spec.rb]: Updated! I went with `thread_class_missing_instrumentation` since I "without" seemed to suggest that we never added it.

**Comment by ivoanjo** [comment]
Note: I force-pushed because I rebased on top of current `feature/profiling` before adding 02d3805.

File: lib/ddtrace/profiling/collectors/stack.rb
Comment (delner) [reply lib/ddtrace/profiling/collectors/stack.rb:127]: I think your logic (all threads share the same ancestors) is sound given our current implementation, but the nature of the check in `get_cpu_time_interval!` is nonetheless assumptive rather than defensive. I think the conditional check might benefit from being a little more fail-safe by being more direct.

It's a bit awkward setting the `@cpu_time_expected_to_work` in `initialize` as an assumed defense of what will take place in `get_cpu_time_interval!`. I would expect such conditions to be evaluated closer to where they're used.

Could the instance variables set in `initialize` instead be bumped inside of `warn_about_missing_cpu_time_instrumentation`, then change `get_cpu_time_interval!`? e.g.

```
def get_cpu_time_interval!(thread)
  # Return if we can't get the current CPU time
  return unless thread.respond_to?(:cpu_time)
  return warn_about_missing_cpu_time_instrumentation(thread) unless thread.respond_to?(:cpu_time)
  # ...
```

I think this might keep things a bit tighter/cleaner by simplifying & isolating the context of each of these functions.

> One extra detail is that you can't even un-include cleanly. I needed to get quite creative in the tests to simulate a case where initialize gets broken, exactly because I couldn't find any other way of skipping our module and only calling the regular initialize. So I'd be struggling to find a case where not every thread is kind_of?(CThread) once one of them is.

If possible, this may be a good time to leverage `expect_in_fork do ... end` which will run the test in the fork, in which you can apply patches without worrying about un-including things.

**Comment by ivoanjo** [comment]
Note: I needed to do another force-push with a rebase on top of latest `feature/profiling`. Long-lived branches :((((((

File: lib/ddtrace/profiling/collectors/stack.rb
Comment (ivoanjo) [reply lib/ddtrace/profiling/collectors/stack.rb:127]: The code suggestion seems like it may be missing something -- with two `unless thread.respond_to?(:cpu_time)` checks, would the warning ever trigger?

Whoever, I think I get the gist of the idea: the concern that we're calling `cpu_time` and `cpu_time_instrumentation_installed?` on a given `thread` object without validating that the object does handle those two methods.

I'm not convinced it's worth being that defensive -- my thinking is that if we're in a really weird corner case, it may just be better for the profiler to stop and report "I died" than to attempt to chug along and ignore any unexpected things. BUT I am prepared to disagree and commit -- you have a lot more experience with this codebase :)

One other thing mentioned was

> It's a bit awkward setting the @cpu_time_expected_to_work in initialize as an assumed defense of what will take place in get_cpu_time_interval!. I would expect such conditions to be evaluated closer to where they're used.

There's a bit more than meets the eye here `@cpu_time_expected_to_work` is being used for two things:
1. It's a neat shortcut for the extra checks that we were previously doing
2. More importantly, and going back to the commit description:
    > This means that regarding CPU instrumentation, we can be in one of
three cases:
    >
    > 1. CPU instrumentation is available AND every thread is correctly
instrumented
     > 2. CPU instrumentation is available BUT there are threads that are
missing the instrumentation
     > 3. CPU instrumentation is NOT available
    
    `@cpu_time_expected_to_work == true` means that we detected that we're in a i. or ii. situation (and not in iii.)

Thus, while we can stop relying on `@cpu_time_expected_to_work` for 1., it is a key part of the detection for 2.

With that said, what do you think of changing `get_cpu_time_interval!` to

```ruby
        def get_cpu_time_interval!(thread)
          # Return if we can't get the current CPU time
          return unless @cpu_time_expected_to_work
          
          unless thread.respond_to(:cpu_time) && thread.cpu_time_instrumentation_installed?
            return warn_about_missing_cpu_time_instrumentation(thread)
          end
```

or

```ruby
        def get_cpu_time_interval!(thread)
          # Return if we can't get the current CPU time
          return unless @cpu_time_expected_to_work
          
          unless thread.respond_to(:cpu_time_instrumentation_installed?) && thread.cpu_time_instrumentation_installed?
            return warn_about_missing_cpu_time_instrumentation(thread)
          end
```

Do you think it would be a reasonable trade-off?

Comment (delner) [reply lib/ddtrace/profiling/collectors/stack.rb:127]: The suggestion made would be a slight improvement on safety, but doesn't really address the awkwardness of `@cpu_time_expected_to_work` being declared in `#initialize`. I appreciate the usefulness of evaluating this value, but why must this be declared and stored as an instance variable? And/or why can't this evaluation be done within `warn_about_missing_cpu_time_instrumentation`?

I think it'd be cleaner/easier to reason about to have this state/evaluation take place in the function its used for, rather than be evaluated piecemeal across multiple different functions at different times, unless there's a performance consideration that isn't insignificant.

Either way, I don't feel strongly enough about this to continue blocking the PR on this point, it was just more a point of feedback. Let me know what you'd prefer to do, and we can move this forward.

**Comment by delner** [comment]
Yeah, it's okay to force-push these branches as long as they aren't shared. We'll get this merged soon!

File: lib/ddtrace/profiling/collectors/stack.rb
Comment (ivoanjo) [reply lib/ddtrace/profiling/collectors/stack.rb:127]: > I appreciate the usefulness of evaluating this value, but why must this be declared and stored as an instance variable? 

In the current version of the code, the `@cpu_time_expected_to_work` check in the initializer is needed to determine if we are in a i. or ii. situation (it's available but we're not sure if all threads are instrumented), not in iii. (it's not available).

If we're on macOS, `@cpu_time_expected_to_work` becomes `false` and thus we return immediately from `get_cpu_time_interval!`.

Without this distinction, we would emit the warning for systems in which the CPU profiling is unsupported, which would be incorrect.

> And/or why can't this evaluation be done within `warn_about_missing_cpu_time_instrumentation`?

I guess something like

```ruby
        def warn_about_missing_cpu_time_instrumentation(thread)
          return if @warned_about_missing_cpu_time_instrumentation

          # make sure we warn only once
          @warned_about_missing_cpu_time_instrumentation = true

          if Thread.current.respond_to?(:cpu_time) && Thread.current.cpu_time
            Datadog.logger.warn(
              "Detected thread ('#{thread}') with missing CPU profiling instrumentation. " \
              'CPU Profiling results will be inaccurate. ' \
              'Please report this at https://github.com/DataDog/dd-trace-rb/blob/master/CONTRIBUTING.md#found-a-bug'
            )
          else
            # no warning -- CPU time instrumentation does not seem to be available
          end

          nil
        end
```

would work -- getting rid of `@cpu_time_expected_to_work` completely. The trade-off is testing becomes a lot more awkward: the current test suite just dependency-injects the value of `cpu_time_expected_to_work` to test the behaviors, without needing to further monkey patch threads.

> Either way, I don't feel strongly enough about this to continue blocking the PR on this point, it was just more a point of feedback. Let me know what you'd prefer to do, and we can move this forward.

I'm open to addressing your concerns 😄 

Between this comment, as well as the one above, I've provided a few alternative approaches and their trade-offs -- could you guide me towards the solution that you prefer?

Comment (delner) [reply lib/ddtrace/profiling/collectors/stack.rb:127]: Granting the nature of this particular component is trickier than some, I don't think dependency injection is always the best strategy to handle awkward tests like these, especially if we're writing code for the sake of tests and not better design.

This point is wrought of experience with our tracing test suite: we had once heavily relied on dependency injection to drive tests for integrations (namely a `:tracer` configuration option), which ultimately resulted in a huge blind spot and a major bug because the tests heavily leveraged the injection, whereas the real applications did not. This is where my mindset was coming from as a point of risk.

As a general rule of thumb, I think its better to leverage the test suite mechanisms wherever reasonable first (e.g. mocks, etc) before designing code around tests. That said, general rules apply generally; it's no substitute for good judgment, so apply your own.

I think the code sample you shared above is simpler and easier to reason about, so my personal preference is for that.

**Review by delner — APPROVED** [review]
Left one more suggestion, but feel free to merge this as is if you see fit! I think this warning will be an excellent feature, so I'll be glad to see this :)

File: lib/ddtrace/profiling/collectors/stack.rb
Comment (ivoanjo) [reply lib/ddtrace/profiling/collectors/stack.rb]: > This point is wrought of experience with our tracing test suite: we had once heavily relied on dependency injection to drive tests for integrations (namely a :tracer configuration option), which ultimately resulted in a huge blind spot and a major bug because the tests heavily leveraged the injection, whereas the real applications did not. This is where my mindset was coming from as a point of risk.

I think there's a balance to be struck, e.g. we need to test it _somewhere_, but that doesn't mean we need to test it everywhere.

Comment (ivoanjo) [reply lib/ddtrace/profiling/collectors/stack.rb]: In c614edc31e7c01dc2ece95187827326f0dd9826e I've removed the use of `@cpu_time_expected_to_work` in `get_cpu_time_interval!` and in fe5db70e55ec189f215d6bf4f5dcbffef16cf7a3 I removed `@cpu_time_expected_to_work` entirely.

**Pull request merged**
